core/src/saros/concurrent/management/ConcurrentDocumentClient.java

```diff
@@ -32,10 +35,23 @@
 
   private final JupiterClient jupiterClient;
 
-  public ConcurrentDocumentClient(ISarosSession sarosSession) {
+  /**
+   * A set of shared resources that were deleted during the session. It is used to filter out
+   * activities for such resources that were created before the other participants ran the
+   * corresponding resource deletion activity locally.
+   *
+   * <p>The set is updated once the resource is recreated as we then want to handle activities for
+   * it again.
+   */
+  // FIXME this could introduce issues when the order of activities is changed, i.e. we receive the
+  //  content change before the file creation
+  private final Set<SPath> deletedResources;
 
+  public ConcurrentDocumentClient(ISarosSession sarosSession) {
     this.sarosSession = sarosSession;
     this.jupiterClient = new JupiterClient(sarosSession);
+
+    this.deletedResources = new CopyOnWriteArraySet<>();
```

Comment:
This is single threaded, both the Client and the Server should only be accessed by a dedicated thread (yes I know the reset stuff does it not correctly)

Reply:
Done. But this only applies to the client. The server does not have the same guarantees according to the javadoc of `transformIncoming(...)`.

Reply:
The client should only be accessed through SWT/AWT/Application Thread. The server (if I am not mistaken) is accessed only the the NetworkDispatchThread (which IMO is pretty bad as it blocks every message send through the IPacketListeners). But I remember there was a dirty hook (client / server ?) that reacts on user leave events and that is main reason why there are concurrent collections / synchronized stuff although it should not.

Reply:
Hmm. But wouldn't it then be the better option to keep it as a synchronized access in both classes and add a todo to make it unsynchronized once we are sure that it's accessed correctly? The overhead from having the set synchronized should be minimal.